holoviews/core/data/grid.py

```diff
@@ -72,16 +84,30 @@ def init(cls, eltype, data, kdims, vdims):
         kdim_names = [d.name if isinstance(d, Dimension) else d for d in kdims]
         vdim_names = [d.name if isinstance(d, Dimension) else d for d in vdims]
         expected = tuple([len(data[kd]) for kd in kdim_names])
+        shapes = tuple([data[kd].shape for kd in kdim_names])
         for vdim in vdim_names:
             shape = data[vdim].shape
             error = DataError if len(shape) > 1 else ValueError
-            if shape != expected[::-1] and not (not expected and shape == (1,)):
+            if (not expected and shape == (1,)) or (len(set((shape,)+shapes)) == 1 and len(shape) > 1):
+                # If empty or an irregular mesh
+                pass
+            elif len(shape) != len(expected):
+                raise error('The shape of the %s value array does not '
+                            'match the expected dimensionality indicated '
+                            'by the key dimensions. Expected %d-D array, '
+                            'found %d-D array.' % (vdim, len(expected), len(shape)))
+            elif any((s!=e and (s+1)!=e) for s, e in zip(shape, expected[::-1])):
                 raise error('Key dimension values and value array %s '
                             'shapes do not match. Expected shape %s, '
                             'actual shape: %s' % (vdim, expected[::-1], shape), cls)
         return data, {'kdims':kdims, 'vdims':vdims}, {}
 
 
+    @classmethod
+    def irregular(cls, dataset, dim):
```

Comment:
You've used the word ``irregular`` a lot. I am wondering if ``uneven`` would have been clearer though it probably isn't worth changing now.

Reply:
irregular != uneven, they are different concepts.